### PR TITLE
Add AuxPoW support

### PR DESCRIPTION
**Pull request by rnicoll**
This introduces AuxPoW support based on the heavily updated code in Namecore. Currently this is only test as far as unit tests, and other changes will no doubt be required, however I wanted to get us all working on the same code base.

### CODE REVIEW

**Comment by patricklodder** [comment]
utACK, we need #1201 to test this properly

File: src/rpcmining.cpp [hunk 1]
```diff
@@ -726,3 +727,189 @@ Value estimatepriority(const Array& params, bool fHelp)
...
+            + HelpExampleCli("getauxblock", "\"hash\" \"serialised auxpow\"")
+            + HelpExampleRpc("getauxblock", "")
+            );
+
+    if (pwalletMain == NULL)
+        throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Method not found (disabled)");
+
+    if (vNodes.empty() && !Params().MineBlocksOnDemand())
+        throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED,
+                           "Namecoin is not connected!");
```
Comment (langerhans) [inline src/rpcmining.cpp]: Just randomly saw this. Should probably be changed as it is user-facing.

File: src/rpcmining.cpp [hunk 2]
```diff
@@ -726,3 +727,189 @@ Value estimatepriority(const Array& params, bool fHelp)
...
+
+    if (IsInitialBlockDownload() && !Params().MineBlocksOnDemand())
+        throw JSONRPCError(RPC_CLIENT_IN_INITIAL_DOWNLOAD,
+                           "Namecoin is downloading blocks...");
```
Comment (langerhans) [inline src/rpcmining.cpp]: Same here

File: src/dogecoin.cpp [hunk 3]
```diff
@@ -75,6 +75,51 @@ unsigned int CalculateDogecoinNextWorkRequired(const CBlockIndex* pindexLast, in
     return bnNew.GetCompact();
 }
 
+bool CheckAuxPowProofOfWork(const CBlockHeader& block, const Consensus::Params& params)
+{
+    /* Except for legacy blocks with full version 1, ensure that
+       the chain ID is correct.  Legacy blocks are not allowed since
+       the merge-mining start, which is checked in AcceptBlockHeader
+       where the height is known.  */
+    if (!block.nVersion.IsLegacy() && params.fStrictChainId && block.nVersion.GetChainId() != params.nAuxpowChainId)
```
Comment (langerhans) [inline src/dogecoin.cpp]: At the moment chainID does not get set on blocks if I read it correctly. At least on regtest, newly generated blocks will have nVersion = 3 so this verification fails on regtest. We should either set the CURRENT_VERSION in CBlock to one with the correct chainID or call SetChainId on the newly created block. getauxblock handles this afaik so this should be merely a regtest issue (testnet has fStrictChainId = false).

File: src/miner.cpp [hunk 4]
```diff
@@ -101,6 +101,9 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
         return NULL;
     CBlock *pblock = &pblocktemplate->block; // pointer for convenience
 
+    /* Initialise the block version.  */
+    pblock->nVersion = CBlockHeader::CURRENT_VERSION;
```
Comment (langerhans) [inline src/miner.cpp]: See here: https://github.com/dogecoin/dogecoin/blob/master/src/core.h#L403 
That's what my other comment was about.

File: src/dogecoin.cpp (hunk 3 above)
Comment (rnicoll) [inline src/dogecoin.cpp]: Somehow https://github.com/namecoin/namecore/blob/master/src/primitives/pureheader.cpp#L12 and calls to it are missing. Must have been in a commit I missed. I'll get that fixed tomorrow.

Comment (rnicoll) [inline src/dogecoin.cpp]: Oh, I see, missed chain ID when in the fancy operator overloading.

File: src/miner.cpp (hunk 4 above)
Comment (rnicoll) [inline src/miner.cpp]: It's now in https://github.com/dogecoin/dogecoin/blob/1.10-dev/src/primitives/pureheader.h#L64 via https://github.com/dogecoin/dogecoin/blob/1.10-dev/src/miner.cpp#L106

**Pull request merged**